server/WFCORE-4251_list_modules_in_deployment_classpath.adoc

```diff
@@ -0,0 +1,118 @@
+= Provide ability to list modules which are on deployment's classpath via a Management Operation.
+:author:            Yeray Borges
+:email:             yborgess@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+:issue-base-url:    https://issues.jboss.org/browse/
+
+== Overview
+Currently, the list of modules available at classpath by each deployment is only available via jconsole:
+
+jboss.modules -> ModuleLoader -> ServiceModuleLoader -> Operations -> getDependencies( "deployment.<application_war_ear_name>" )
+
+This document describes the RFE to provide the ability to expose this information via a management operation.
+
+When optimizing the application footprint, this information becomes important:
+
+* The user could use this list to identify which dependencies are added implicitly by the server, and then, exclude any unneeded dependency from the deployment.
+* The user could use this information to discover any duplicated artifact bundled together with the deployment and included in one of the modules that the server is adding to the application classpath.
+* This information could facilitate discovering unneeded subsystems in the installation of his servers.
+
+
+The proposed solution is to include a new operation `list-modules` available under `/deployment` and, in case of ears, `/deployment=*/subdeployment` resource:
+
+ [standalone@localhost:9990 /] /deployment=ejb-in-ear.ear:list-modules
+
+ [standalone@localhost:9990 /] /deployment=ejb-in-ear.ear/subdeployment=ejb-in-ear-web.war:list-modules
+
+An example of the information returned by this method could be:
+
+  {
+      "outcome" => "success",
+      "result" => {
+          "ejb-in-ear.ear" => [
+              {
+                  "name" => "com.fasterxml.jackson.datatype.jackson-datatype-jdk8",
+                  "slot" => "main",
+                  "optional" => true,
+                  "export" => false,
+                  "import-services" => true,
+                  "duplicated" => false
+              },
+              {
+                  "name" => "com.fasterxml.jackson.datatype.jackson-datatype-jsr310",
+                  "slot" => "main",
+                  "optional" => true,
+                  "export" => false,
+                  "import-services" => true,
+                  "duplicated" => true
+              },
+              ...
```

Comment:
Does this list include the deployment module itself -- i.e. the one created for it by WildFly?

Reply:
Yes, it does.